### PR TITLE
Add a HIP kernel to construct an array of batched pointers.

### PR DESCRIPTION
**Pull request by rsanthanam-amd**

### CODE REVIEW

**Comment by rsanthanam-amd** [comment]
/cc @jayfurmanek @weihanmines 

To use this add the following to your source file:

#include "tensorflow/stream_executor/gpu/gpu_stream.h"

namespace stream_executor {
namespace gpu {

extern void rocm_MakeBatchPointers(void* stream, char* base, int stride, int n, void** ptrs_out);

}}

Here is an example call:

  stream_executor::gpu::rocm_MakeBatchPointers(se::gpu::AsGpuStreamValue(stream),
                         reinterpret_cast<char*>(a_base),
                         stride_bytes, batch_size_,
                         reinterpret_cast<void**>(as.opaque()));

**Comment by rsanthanam-amd** [comment]
@ekuznetsov139 please check the launch parameters in the hipLaunchKernelGGL - not sure if i got that quite right.

**Comment by rsanthanam-amd** [comment]
@ekuznetsov139 please check the launch parameters in the hipLaunchkernelGGL call - not sure if i got that quite right.

File: tensorflow/stream_executor/rocm/rocm_helpers.cu.cc [hunk 1]
```diff
@@ -31,5 +31,16 @@ void broadcast_fp32(void* stream, float* dst, int dst_stride, int batches, int s
...
                      (hipStream_t)stream, dst, dst_stride, batches, src, size);
 }
+__global__ void __xla_MakeBatchPointers(char* base, int stride, int n, void** ptrs_out) {
+  int idx = threadIdx.x + blockIdx.x * blockDim.x;
+  if (idx >= n) return;
+  ptrs_out[idx] = base + idx * stride;
+}
+
+void rocm_MakeBatchPointers(void* stream, char* base, int stride, int n, void** ptrs_out) {
+  hipLaunchKernelGGL(__xla_MakeBatchPointers, dim3((n + 256 - 1)/256, 1, 1), dim3(256, 1, 1), 0,
```
Comment (deven-amd) [inline tensorflow/stream_executor/rocm/rocm_helpers.cu.cc]: explain what 256 is....change this to something like

```
int threads_per_block = 256;
int num_blocks = (batch_size + threads_per_block - 1) / threads_per_block;
```

File: tensorflow/stream_executor/rocm/rocm_helpers.cu.cc [hunk 2]
```diff
@@ -31,5 +31,16 @@ void broadcast_fp32(void* stream, float* dst, int dst_stride, int batches, int s
   hipLaunchKernelGGL(broadcast_fp32_kernel, dim3(x_blocks, (batches+3)/4, src_batches), min(256, (int)size), 0,
                      (hipStream_t)stream, dst, dst_stride, batches, src, size);
 }
+__global__ void __xla_MakeBatchPointers(char* base, int stride, int n, void** ptrs_out) {
+  int idx = threadIdx.x + blockIdx.x * blockDim.x;
+  if (idx >= n) return;
+  ptrs_out[idx] = base + idx * stride;
+}
+
```
Comment (deven-amd) [inline tensorflow/stream_executor/rocm/rocm_helpers.cu.cc:44]: add a comment block here explaining what this routine does

**Review by deven-amd — COMMENTED** [review]
can we append the PR to add a call to this routine as well?

**Comment by jayfurmanek** [comment]
Trying to use this for triangular solve.
Getting a fail in the kernel so there's a good chance I'm not using it correctly.

@rsanthanam-amd  can you take a quick look over: (thx!)
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/pull/1589

The CUDA version uses `DeviceMemoryBase` objects, which looks like:
```
Status MakeBatchPointers(se::Stream* stream, const se::GpuAsmOpts& asm_opts,
                         se::DeviceMemoryBase base_ptr, int stride_bytes, int n,
                         se::DeviceMemoryBase ptrs_out)
```
pulling opaque ptrs out of there should work, so not sure what I'm doing wrong.

**Comment by jayfurmanek** [comment]
I guess that's one review comment. Can/should we keep the function signature of the helper the same as the CUDA version so that it's essentially a drop in replacement? The helper should be able to get what it needs from those DeviceMemoryBase objects.

**Comment by rsanthanam-amd** [comment]
> I guess that's one review comment. Can/should we keep the function signature of the helper the same as the CUDA version so that it's essentially a drop in replacement? The helper should be able to get what it needs from those DeviceMemoryBase objects.

I would prefer to keep the function arguments as fundamental data types in this file and not force a dependence on DeviceMemoryBase objects.

**Comment by rsanthanam-amd** [comment]
> can we append the PR to add a call to this routine as well?

Unfortunately, my use case PR is not ready yet.

File: tensorflow/stream_executor/rocm/rocm_helpers.cu.cc (hunk 1 above)
Comment (rsanthanam-amd) [reply tensorflow/stream_executor/rocm/rocm_helpers.cu.cc]: done.

File: tensorflow/stream_executor/rocm/rocm_helpers.cu.cc (hunk 2 above)
Comment (rsanthanam-amd) [reply tensorflow/stream_executor/rocm/rocm_helpers.cu.cc:44]: done.

**Comment by rsanthanam-amd** [comment]
Modified the existing MakeBatchPointers call to include the rocm variant.

Now both CUDA and ROCm users can be serviced with a single function call.

**Comment by rsanthanam-amd** [comment]
The CUDA implementation of MakeBatchPointers is more sophisticated since it caches a kernel instance for every stream executor.

Not sure if it makes sense to spend time on doing the same for ROCm.

Any thoughts?

**Comment by rsanthanam-amd** [comment]
Perhaps the caching scheme for CUDA exists to avoid JITing the PTX multiple times for the same stream executor.

If that is the only reason, then I think we are probably ok.

**Pull request merged**